cores/arduino/USBCore.cpp

```diff
@@ -624,10 +625,51 @@ int USBCore_::send(uint8_t ep, const void* data, int len)
     auto flags = ep & 0xf0;
     ep &= 0x7;
     auto wrote = 0;
+    auto usbd = &USBCore().usbDev();
 
+#ifdef USBD_REMOTE_WAKEUP
+    /*
+     * The workaround below is for a complex and subtle bug. The
+     * low-level USBD firmware tracks the USBD peripheral state in
+     * usbd->cur_status. The suspend interrupt handler saves it in
+     * usbd->backup_status, and then sets it to USBD_SUSPENDED. The
+     * wakeup interrupt handler restores it.
+     *
+     * Due to a limitation or bug in the USBD peripheral, the suspend
+     * interrupt can fire multiple times if the application sends a
+     * remote wakeup. This can overwrite usbd->backup_status, causing
+     * the wakeup interrupt to incorrectly restore the status as
+     * USBD_SUSPENDED. The host and USBD peripheral will both think
+     * the device is resumed, but the low-level firmware thinks it's
+     * still suspended, and won't respond appropriately to bus traffic
+     * or application requests.
+     *
+     * The workaround restores usbd->cur_status before the suspend
+     * handler erroneously overwrites the backup. The
+     * standard_hid_keyboard example in the GD32F30x firmware bundle
+     * uses this workaround, without explanation.
+     *
+     * Application note AN077 describes possible multiple firings of
+     * the suspend interrupt during a remote wakeup request, but also
+     * doesn't adequately explain. AN077 also implies that ESOFIF is
+     * set on any missing SOF, disregarding whether the line state is
+     * suspend (long-"J-like") or resume (long-K). It also implies
+     * that SPSIF similarly does not distinguish these line states.
+     */
```

Comment:
the following check-and-set should be wrapped in `usb_{en,dis}able_interrupts` to make sure it doesn't get clobbered on the next sof

Reply:
The example code for the GD32 USBD firmware doesn't disable interrupts around this. Also, the SOF interrupt handler itself doesn't touch `cur_status` or `backup_status`. It might be safe to reverse the order, but it seems unlikely that the suspend interrupt will fire again (3ms) between those two statements.

A more robust fix might be to change the suspend handler in the low-level driver to only save `cur_status` if it's not already `USBD_SUSPENDED`. I was trying to avoid touching that code, if possible.

Reply:
my reading of `usbd_isr` is that when the `SPSIF` flag in `USBD_INTF` is set (which is triggered when 3ms go by without a sof), both `backup_status` and `cur_status` are set.

i agree the odds are miniscule, but the cost of wrapping it is, too. in addition, there are other places from interrupt context that can write to `cur_status`, such as during enumeration, although, again, that's probably pretty unlikely (if not outright impossible).

indeed, the ideal place would be to fix it in the firmware library itself, but we'll need to go through other channels for that. there's another bug around buffers only being written to 2 bytes at a time that should also be upstreamed as well.

Reply:
I'm concerned that the interrupt handlers in `usbd_lld_int.c` might manipulate the interrupt enable states. I'll need to check that.

Reply:
Pushed an update adding interrupt blocking in that section.